File: opabsl/mlbsl/bslString.ml
(*
    Copyright © 2011, 2012 MLstate

    This file is part of OPA.

    OPA is free software: you can redistribute it and/or modify it under the
    terms of the GNU Affero General Public License, version 3, as published by
    the Free Software Foundation.

    OPA is distributed in the hope that it will be useful, but WITHOUT ANY
    WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
    FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for
    more details.

    You should have received a copy of the GNU Affero General Public License
    along with OPA. If not, see <http://www.gnu.org/licenses/>.
*)
##register concat \ `Pervasives.( ^ )` : string, string -> string

##register length \ `String.length` : string -> int

##register get : string, int -> string
  let get s i = String.make 1 (String.get s i)

##register repeat \ `BaseString.repeat` : string, int -> string

##register sub : int, int, string -> string
  let sub start len src = String.sub src start len

##register replace : string, string, string -> string
  let replace search replacement source =
    BaseString.replace source search replacement

##register strip_quotes \ `BaseString.strip_quotes` : string -> string

##register index : string, string -> option(int)
  let index pattern source =
    let lp = String.length pattern in
    let ls = String.length source in
    let rec aux id =
      if ls - lp - id >= 0 then
        if BaseString.is_substring pattern source id then Some id
        else aux (id + 1)
      else
        None
    in
      aux 0

(*Low-level: fails with UTF-8*)
##register reverse : string -> string
  let reverse s =
    let len = String.length s in
    let res = String.make len ' ' in
    let i = ref 1 in
      String.iter (fun c ->
                   String.set res (len - !i) c;
                   i := !i + 1;
                   ()) s;
      res

##register lowercase \ `Cactutf.lowercase` : string -> string

##register uppercase \ `Cactutf.uppercase` : string -> string

##register remove_accents \ `BaseString.remove_accents` : string -> string

let have_to_be_escaped_table =
  let have_to_be_escaped chr =
    let code = Char.code chr in
    code >= 128 || String.contains Base.Utf8.except_html_char chr || (code < 32 && not (String.contains Base.Utf8.allowed_special_char chr)) in
  Array.init 256 (fun code -> have_to_be_escaped (Char.unsafe_chr code))
let have_to_be_escaped (c:char) = Array.unsafe_get have_to_be_escaped_table (Char.code c)
let not_have_to_be_escaped (c:char) = not (have_to_be_escaped c)

(* too slow
let _utf8_byte_have_to_be_escaped = function
  | '"' | '<' | '>' | '&' -> true
(* | '\'' -> true *)
  | _ -> false
*)
(* Don't understand how using an array is faster in this case, but it is *)
let table_utf8_byte_have_to_be_escaped = Array.init 256 (fun code -> String.contains Base.Utf8.except_html_char (Char.unsafe_chr code))
let utf8_byte_have_to_be_escaped c = Array.unsafe_get table_utf8_byte_have_to_be_escaped (Char.code c)


let utf8_byte_not_have_to_be_escaped c = not (utf8_byte_have_to_be_escaped c)

(* fast path for non escaped part x2 speed-up *)
(* same as BaseString.len_from but avoiding second order call and hopefully benefit from inlining *)
let rec from_to_utf8_nbhtbe src i len =
  if (i<>len) && (utf8_byte_not_have_to_be_escaped (String.unsafe_get src i)) then from_to_utf8_nbhtbe src (i+1) len else i
let rec from_to_nhtbe src i len =
  if (i<>len) && (not_have_to_be_escaped (String.unsafe_get src i)) then from_to_nhtbe src (i+1) len else i

let empty_buf = Buffer.create 0

(* This thing works with utf-8 because
   - if utf8 encoding is ok, no 'one byte utf8 char' needs to be escaped if the html has utf-8 encoding,
   - if utf8 is not ok, all byte of longer than on byte character are seen as needing escaping *)
##register escapeHTML : bool, string -> string
let escapeHTML utf8 src =
  let len = String.length src in
  if len=0 then "" else
  let have_to_be_escaped     = if utf8 then utf8_byte_have_to_be_escaped     else have_to_be_escaped     in
  let not_have_to_be_escaped = if utf8 then utf8_byte_not_have_to_be_escaped else not_have_to_be_escaped in
  let rec aux pos buf =
    let need_escaping = have_to_be_escaped src.[pos] in
    let len_to_push =
      if utf8 && not(need_escaping) then (from_to_utf8_nbhtbe src pos len)-pos
      else if not(need_escaping) then (from_to_nhtbe src pos len)-pos
      else (
        let case = if need_escaping then have_to_be_escaped else not_have_to_be_escaped in
        BaseString.len_from case src pos
      )
    in
    let new_pos = pos + len_to_push in
    if new_pos >= len && buf == empty_buf then (* optimize for the escape-in-one-step case *)
      (if need_escaping then Base.Utf8.htmlentities src else src)
    else
      let buf = if buf == empty_buf then Buffer.create (len+(4*len/(new_pos+1))) else buf in
      if need_escaping then
        Base.Utf8.htmlentities_append buf src pos len_to_push
      else
        Buffer.add_substring buf src pos len_to_push;
      if new_pos < len then
        aux new_pos buf
      else
        Buffer.contents buf
  in
  aux 0 empty_buf

##register to_character \ `Base.Utf8.string_of_int` : int -> string

##register of_int \ `Pervasives.string_of_int` : int -> string

##register of_byte_val : int -> string
let of_byte_val byte =
  try
    String.make 1 (Char.chr byte)
  with
  | Invalid_argument _ -> "\000"

##register of_byte_unsafe : int -> string
let of_byte_unsafe i =
  String.make 1 (Base.Char.chr i)

##register byte_at_unsafe : int, string -> int
let byte_at_unsafe n s = Base.Char.code s.[n]
   (* special function for TRX *)
   (* TODO write it in C for better performance (on pointers)?
           we could then even use some bit-level magic cleverlness to compare word-by-word instead
           of byte-by-byte*)
##register check_match_literal : string, int, string -> bool
  let check_match_literal input pos literal =
    let n = String.length literal in
    let i = ref 0 in
    while !i < n && String.unsafe_get input (pos + !i) == String.unsafe_get literal !i do
      incr i
    done;
    !i == n


##register leq: string, string -> bool
let leq (a:string) (b:string) = a <= b

##register lt: string, string -> bool
let lt (a:string) (b:string) = a < b

##register eq: string, string -> bool
let eq (a:string) (b:string) = a = b

##register geq: string, string -> bool
let geq (a:string) (b:string) = a >= b

##register gt: string, string -> bool
let gt (a:string) (b:string) = a > b

##register neq: string, string -> bool
let neq (a:string) (b:string) = a <> b

##register ordering: string, string -> opa[Order.ordering]
let ordering (a:string) (b:string) =
  match String.compare a b with
  | -1 -> BslPervasives.ord_result_lt
  | 0 -> BslPervasives.ord_result_eq
  | 1 -> BslPervasives.ord_result_gt
  | _ -> assert false

##register encode_uri_component\ `Encodings.encode_uri_component`: string -> string
##register decode_uri_component\ `Encodings.decode_uri_component`: string -> string

File: opabsl/jsbsl/bslString.js
/*
    Copyright © 2011, 2012 MLstate

    This file is part of OPA.

    OPA is free software: you can redistribute it and/or modify it under the
    terms of the GNU Affero General Public License, version 3, as published by
    the Free Software Foundation.

    OPA is distributed in the hope that it will be useful, but WITHOUT ANY
    WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
    FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for
    more details.

    You should have received a copy of the GNU Affero General Public License
    along with OPA.  If not, see <http://www.gnu.org/licenses/>.
*/

/**
 * Low-level string operations
 *
 * To access these operations at OPA-level, use module [String].
 */
##register concat: string, string -> string
##args(a, b)
{
    return a + b;
}

/**
 * Replace all occurrences of a substring in a given string
 *
 * @param search
 * @param replacement
 * @param source
 */
##register replace : string, string, string -> string
  ##args(search, replacement, source)
   {
    var left= "", right= source, pos;
    while (true)
    {
      pos= right.indexOf(search);
      if (pos < 0) return left+right;
      left += right.substr(0, pos)+replacement;
      right= right.substr(pos + search.length);
    }
   }

##register repeat : string, int -> string
##args(s, n)
{
    for (var str= "", i= 0; i < n; i++)
        str += s ;
    return str;
  }

##register get : string, int -> string
  ##args(s, i)
  {
    return s[i];
  }

##register reverse : string -> string
  ##args(s)
  {
    return s.split('').reverse().join('');
  }

##register length : string -> int
  ##args(s)
  {
    return s.length;
  }

##register sub : int, int, string -> string
  ##args(offset, len, src)
  {
    return src.substr(offset, len);
  }

##register index : string, string -> option(int)
  ##args(pattern, source)
  {
    var pos = source.indexOf(pattern);
    return (pos > -1) ? js_some(pos) : js_none;
  }

##register uppercase : string -> string
  ##args(str)
  {
    return str.toUpperCase();
  }

##register lowercase : string -> string
##args(str)
{
    return str.toLowerCase();
}

##register remove_accents: string -> string
##args(str)
{
    var result = str.replace(/[ÀÁÂÃÄÅ]/g, "A")
        .replace(/[àáâãäå]/g, "a")
        .replace(/[ÒÓÔÕÖØ]/g, "O")
        .replace(/[òóôõöø]/g, "o")
        .replace(/[ÈÉÊË]/g,   "E")
        .replace(/[èéêë]/g,   "e")
        .replace(/Ç/g,        "C")
        .replace(/ç/g,        "c")
        .replace(/ÌÍÎÏ/g,     "I")
        .replace(/ìíîï/g,     "i")
        .replace(/ÙÚÛÜ/g,     "u")
        .replace(/ùúûü/g,     "u")
        .replace(/ÿ/g,        "y")
        .replace(/Ñ/g,        "N")
        .replace(/ñ/g,        "n")
    return result;
}

##register escapeHTML : bool, string -> string
  ##args(_,someText)
  {
    var div = document.createElement('div');
    var text = document.createTextNode(someText);
    div.appendChild(text);
    return div.innerHTML;
  }

##register of_int : int -> string
  ##args(n)
  {
    return ""+n;
  }

##register to_character : int -> string
  ##args(i)
  {
    /* [String.fromCharCode] doesn't support characters beyond 65535.
     * The following workaround comes from:
     * https://developer.mozilla.org/en/Core_JavaScript_1.5_Reference/Global_Objects/String/fromCharCode
     */

    if (i > 0xFFFF) {
      i -= 0x10000;
      return String.fromCharCode(0xD800 + (i >> 10), 0xDC00 +
                                 (i & 0x3FF));
    }
    else {
      return String.fromCharCode(i);
    }
  }

##register of_byte_unsafe : int -> string
  ##args(i)
  {
    return String.fromCharCode(i);
  }

##register byte_at_unsafe : int, string -> int
  ##args(n, s)
  {
    return s.charCodeAt(n);
  }

   // special function for TRX
##register check_match_literal : string, int, string -> bool
  ##args(input, pos, literal)
  {
    var i = 0;
    for (i = 0; i < literal.length; i++)
        if (input[pos+i] != literal[i])
            return false;
    return true;
  }


function string_escaped(s){ return s; } // what's wrong ?
function string_encode(str){ return escape(str); }
function string_decode(str){ return unescape(str); }


##register leq:string, string -> bool
##args(c1,c2)
{
   return c1 <= c2
}

##register lt:string, string -> bool
##args(c1,c2)
{
   return c1 < c2
}

##register eq:string, string -> bool
##args(c1,c2)
{
   return c1 == c2
}

##register geq:string, string -> bool
##args(c1,c2)
{
   return c1 >= c2
}

##register gt:string, string -> bool
##args(c1,c2)
{
   return c1 > c2
}

##register neq:string, string -> bool
##args(c1,c2)
{
   return c1 != c2
}

##register ordering: string,string -> opa[Order.ordering]
##args(c1,c2)
{
    if(c1<c2) return %%BslPervasives.order_lt%%
    if(c1==c2) return %%BslPervasives.order_eq%%
    return %%BslPervasives.order_gt%%
}

##register encode_uri_component\ encodeURIComponent: string -> string

